### PR TITLE
add test to ensure mqa and mha have the same behaviour

### PR DESCRIPTION
**Pull request by minimario**
- Added a test to compare MHA, MQA1, and MQA2
- Fixed small bug

### CODE REVIEW

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py [hunk 1]
```diff
@@ -134,7 +134,7 @@ def __init__(self, config, is_cross_attention=False, layer_idx=None):
         self.embed_dim = config.hidden_size
         self.num_heads = config.num_attention_heads
         self.head_dim = self.embed_dim // self.num_heads
-        self.kv_heads = 1 if self.is_mqa else self.head_dim
+        self.kv_heads = 1 if self.is_mqa else self.num_heads
```
Comment (jlamypoirier) [inline src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:137]: Good find!

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py [hunk 2]
```diff
@@ -807,3 +815,79 @@ def test_contrastive_search_gpt_bigcode(self):
...
+class GPTBigCodeAttentionTest(unittest.TestCase):
+    def get_attention(self, attention_type : AttentionType):
+        config = GPTBigCodeConfig.from_pretrained("bigcode/santacoder-fast-inference")
+        config.attention_type = attention_type
+        return GPTBigCodeAttention(config)
+
+    def test_mqa_correctness(self):
+        embed_dim = 2048
+        head_dim = 128
+        random_attn_weight = torch.randn(embed_dim, embed_dim)
```
Comment (jlamypoirier) [inline tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:829]: You can simplify these a lot by getting the state dict from one of the MQA attns (it's already initialized), modifying and using `load_state_dict` on the others.

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py [hunk 3]
```diff
@@ -807,3 +815,79 @@ def test_contrastive_search_gpt_bigcode(self):
...
+
+        num_tokens = 5
+        hidden_states = torch.randn(1, num_tokens, embed_dim)
+        attention_mh_result = attention_mh(hidden_states)[0]
+        attention_mq1_result = attention_mq1(hidden_states)[0]
+        attention_mq2_result = attention_mq2(hidden_states)[0]
+
+        self.assertTrue(torch.allclose(attention_mh_result, attention_mq1_result))
+        self.assertTrue(torch.allclose(attention_mh_result, attention_mq2_result))
+        self.assertTrue(torch.allclose(attention_mq1_result, attention_mq2_result))
```
Comment (jlamypoirier) [inline tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:893]: Missing newline, please run `make style`

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py [hunk 4]
```diff
@@ -807,3 +815,79 @@ def test_contrastive_search_gpt_bigcode(self):
...
+        state_dict_mq2["q_attn.weight"] = random_attn_weight
+        state_dict_mq2["q_attn.bias"] = random_attn_bias
+        state_dict_mq2["kv_attn.weight"] = torch.hstack([random_attn_k_weight, random_attn_v_weight])
+        state_dict_mq2["kv_attn.bias"] = torch.hstack([random_attn_k_bias, random_attn_v_bias])
+        state_dict_mq2["c_proj.weight"] = random_proj
+        state_dict_mq2["c_proj.bias"] = random_proj_bias
+        attention_mq2.load_state_dict(state_dict_mq2)
+
+        # Run correctness test
+        attention_mh.eval()
```
Comment (jlamypoirier) [inline tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:881]: Maybe we want to try the training case too?

**Review by jlamypoirier — COMMENTED** [review]
A few suggestions, otherwise looks great. Did you make sure the test passes?

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py [hunk 5]
```diff
@@ -807,3 +811,102 @@ def test_contrastive_search_gpt_bigcode(self):
                 "but said in a statement to The Associated Press that"
             ],
         )
+
+
+@require_torch
+class GPTBigCodeAttentionTest(unittest.TestCase):
+    def get_attention(self, attention_type: AttentionType):
+        config = GPTBigCodeConfig.from_pretrained("bigcode/santacoder-fast-inference")
```
Comment (jlamypoirier) [inline tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:819]: You can disable the dropout here with `attn_pdrop=0`. Also you can set `attn_pdrop` and `attention_type`directly in the `from_pretrained` kwargs.

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py [hunk 6]
```diff
@@ -807,3 +811,102 @@ def test_contrastive_search_gpt_bigcode(self):
...
+            attention_mq1_result = attention_mq1(hidden_states)[0]
+            attention_mq2_result = attention_mq2(hidden_states)[0]
+
+            tolerance = 1e-6
+            self.assertTrue(torch.allclose(attention_mh_result, attention_mq1_result, atol=tolerance))
+            self.assertTrue(torch.allclose(attention_mh_result, attention_mq2_result, atol=tolerance))
+            self.assertTrue(torch.allclose(attention_mq1_result, attention_mq2_result, atol=tolerance))
+
+        # RUN CORRECTNESS TEST IN TRAIN MODE
+        attention_mh.train()
```
Comment (jlamypoirier) [inline tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:886]: This can be merged with eval in a loop.

Comment (jlamypoirier) [reply tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:886]: Or better with a parametrized test (so de know which one failed)

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py [hunk 7]
```diff
@@ -807,3 +811,102 @@ def test_contrastive_search_gpt_bigcode(self):
...
+        attention_mq2.load_state_dict(state_dict_mq2)
+
+        # RUN CORRECTNESS TEST IN EVAL MODE
+        attention_mh.eval()
+        attention_mq1.eval()
+        attention_mq2.eval()
+
+        num_tokens = 5
+
+        for i in range(5):
```
Comment (jlamypoirier) [inline tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:874]: Not sure this loop is useful, it doesn't test with different model weights. Can drop, or use the seed as a test parameter?

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py [hunk 8]
```diff
@@ -807,3 +811,102 @@ def test_contrastive_search_gpt_bigcode(self):
...
+            hidden_states = torch.randn(1, num_tokens, embed_dim)
+            attention_mh_result = attention_mh(hidden_states)[0]
+            attention_mq1_result = attention_mq1(hidden_states)[0]
+            attention_mq2_result = attention_mq2(hidden_states)[0]
+
+            tolerance = 1e-6
+            self.assertTrue(torch.allclose(attention_mh_result, attention_mq1_result, atol=tolerance))
+            self.assertTrue(torch.allclose(attention_mh_result, attention_mq2_result, atol=tolerance))
+            self.assertTrue(torch.allclose(attention_mq1_result, attention_mq2_result, atol=tolerance))
```
Comment (jlamypoirier) [inline tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:912]: Please run make style again

File: src/transformers/models/gpt_bigcode/configuration_gpt_bigcode.py [hunk 9]
```diff
@@ -32,7 +32,7 @@
 }
 
 
-class AttentionType(Enum):
+class AttentionType(IntEnum):
```
Comment (jlamypoirier) [inline src/transformers/models/gpt_bigcode/configuration_gpt_bigcode.py:35]: Why this change?

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py [hunk 10]
```diff
@@ -807,3 +811,88 @@ def test_contrastive_search_gpt_bigcode(self):
...
+        state_dict_mq2["q_attn.weight"] = attn_weight
+        state_dict_mq2["q_attn.bias"] = attn_bias
+        state_dict_mq2["kv_attn.weight"] = torch.hstack([attn_k_weight, attn_v_weight])
+        state_dict_mq2["kv_attn.bias"] = torch.hstack([attn_k_bias, attn_v_bias])
+        state_dict_mq2["c_proj.weight"] = proj
+        state_dict_mq2["c_proj.bias"] = proj_bias
+        attention_mq2.load_state_dict(state_dict_mq2)
+
+        # PUT THE MODEL INTO THE CORRECT MODE
+        if test_mode == "eval":
```
Comment (jlamypoirier) [inline tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:868]: You can just use a bool and set `attention_mh.train(train)`, etc.

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py [hunk 11]
```diff
@@ -807,3 +811,88 @@ def test_contrastive_search_gpt_bigcode(self):
...
+        attention_mq2_result = attention_mq2(hidden_states)[0]
+
+        # CHECK THAT ALL OUTPUTS ARE THE SAME
+        tolerance = 1e-5
+        self.assertTrue(torch.allclose(attention_mh_result, attention_mq1_result, atol=tolerance))
+        self.assertTrue(torch.allclose(attention_mh_result, attention_mq2_result, atol=tolerance))
+        self.assertTrue(torch.allclose(attention_mq1_result, attention_mq2_result, atol=tolerance))
+
+    def test_mqa_correctness_train(self):
+        for seed in range(5):
```
Comment (jlamypoirier) [inline tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:893]: That works, but pytest parametrization would make it easier to know which seed failed. https://docs.pytest.org/en/6.2.x/parametrize.html

**Review by jlamypoirier — APPROVED** [review]
Some minor comments, otherwise this looks great!

File: src/transformers/models/gpt_bigcode/configuration_gpt_bigcode.py (hunk 9 above)
Comment (minimario) [reply src/transformers/models/gpt_bigcode/configuration_gpt_bigcode.py:35]: Without this change, from_pretrained does not work because Enum types cannot be JSON serialized

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py (hunk 11 above)
Comment (minimario) [reply tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py]: I went with parameterized instead because pytest parameterization doesn't seem to play too well with unittest.TestCase

File: src/transformers/models/gpt_bigcode/configuration_gpt_bigcode.py (hunk 9 above)
Comment (jlamypoirier) [reply src/transformers/models/gpt_bigcode/configuration_gpt_bigcode.py:35]: It should be working without the change because the config takes an int, not an AttentionType (line 199). I guess your solution is better though...

Comment (mayank31398) [reply src/transformers/models/gpt_bigcode/configuration_gpt_bigcode.py:35]: agreed

File: tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py [hunk 12]
```diff
@@ -807,3 +812,74 @@ def test_contrastive_search_gpt_bigcode(self):
...
+    def get_attention(self, attention_type: AttentionType):
+        config = GPTBigCodeConfig.from_pretrained(
+            "bigcode/santacoder-fast-inference",
+            attention_type=attention_type,
+            attn_pdrop=0,
+            resid_pdrop=0,
+        )
+        return GPTBigCodeAttention(config)
+
+    @parameterized.expand([(seed, is_train_mode) for seed in range(5) for is_train_mode in [True, False]])
```
Comment (mayank31398) [inline tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py:828]:
```python
@parameterized.expand(list(itertools.product(range(5), [True, False])))
```
this can be written this way. let me know your thoughts.
